File: apps/android/core/data/src/main/kotlin/com/sigmundgranaas/turbo/expressive/core/data/TrailSearchRepository.kt
package com.sigmundgranaas.turbo.expressive.core.data

import com.sigmundgranaas.turbo.expressive.core.common.Outcome
import com.sigmundgranaas.turbo.expressive.domain.LatLng
import com.sigmundgranaas.turbo.expressive.domain.SearchHit
import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.statement.bodyAsText
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import javax.inject.Inject

/**
 * Named-trail search over Geonorge's friluftsruter2 WFS (the Nasjonalturbase
 * foot-route layer). Mirrors the Flutter `TrailSearchService`: a CQL `ILIKE`
 * filter on the route name, surfacing the first vertex of each route as its
 * position so it can be shown as a search hit.
 */
interface TrailSearchRepository {
    suspend fun search(query: String): Outcome<List<SearchHit>>
}

class GeonorgeTrailSearchRepository @Inject constructor(
    private val client: HttpClient,
) : TrailSearchRepository {

    override suspend fun search(query: String): Outcome<List<SearchHit>> = try {
        val escaped = query.trim().replace("'", "''")
        val raw = client
            .get("https://wfs.geonorge.no/skwms1/wfs.friluftsruter2") {
                parameter("SERVICE", "WFS")
                parameter("VERSION", "2.0.0")
                parameter("REQUEST", "GetFeature")
                parameter("TYPENAMES", "fotrute")
                parameter("OUTPUTFORMAT", "application/json")
                parameter("SRSNAME", "urn:ogc:def:crs:EPSG::4326")
                parameter("COUNT", 10)
                parameter("CQL_FILTER", "navn ILIKE '%$escaped%'")
                header("User-Agent", USER_AGENT)
            }
            .bodyAsText()
        Outcome.Success(TrailWfs.parse(raw))
    } catch (t: Throwable) {
        Outcome.Failure(t)
    }

    private companion object {
        const val USER_AGENT = "turbo-expressive/0.1 github.com/SigmundGranaas/turbo"
    }
}

/** Pure GeoJSON-FeatureCollection → [SearchHit] parsing, isolated for testability. */
internal object TrailWfs {
    private val json = Json { ignoreUnknownKeys = true }

    fun parse(body: String): List<SearchHit> =
        runCatching { json.decodeFromString<TrailFeatureCollection>(body) }
            .getOrNull()
            ?.features
            ?.mapNotNull(TrailFeature::toHit)
            .orEmpty()
}

@Serializable
private data class TrailFeatureCollection(val features: List<TrailFeature> = emptyList())

@Serializable
private data class TrailFeature(
    val properties: TrailProperties? = null,
    val geometry: TrailGeometry? = null,
)

@Serializable
private data class TrailProperties(
    val navn: String? = null,
    val rutenummer: String? = null,
    val merkemetode: String? = null,
)

@Serializable
private data class TrailGeometry(
    val type: String? = null,
    val coordinates: JsonElement? = null,
)

private fun TrailFeature.toHit(): SearchHit? {
    val name = properties?.navn?.takeIf(String::isNotBlank) ?: return null
    val firstVertex = geometry?.coordinates?.let(::firstLngLat) ?: return null
    val desc = listOfNotNull(
        "Trail",
        properties.rutenummer?.takeIf(String::isNotBlank),
        properties.merkemetode?.takeIf(String::isNotBlank),
    ).joinToString(" · ")
    return SearchHit(name = name, description = desc, position = firstVertex)
}

/**
 * GeoJSON coordinates are [lng, lat] and arbitrarily nested (Point → pair,
 * LineString → list of pairs, MultiLineString → list of lists). Descend into the
 * first child until a `[lng, lat]` numeric pair is reached, then swap to [LatLng].
 */
private fun firstLngLat(element: JsonElement): LatLng? {
    var node: JsonElement = element
    repeat(MAX_DEPTH) {
        val arr = node as? JsonArray ?: return null
        val head = arr.firstOrNull() ?: return null
        if (head is JsonPrimitive) {
            val lng = (arr.getOrNull(0) as? JsonPrimitive)?.doubleOrNull ?: return null
            val lat = (arr.getOrNull(1) as? JsonPrimitive)?.doubleOrNull ?: return null
            return LatLng(lat, lng)
        }
        node = head
    }
    return null
}

private const val MAX_DEPTH = 6

File: apps/ios/Sources/CoreData/TrailSearchRepository.swift
import Foundation
import CoreModel

/// Search source backed by Geonorge's "Nasjonal turbase" WFS feed (friluftsruter2).
/// Mirrors Flutter `TrailSearchService`. Queries the WFS GetFeature endpoint with
/// a CQL `ILIKE` filter on the trail `navn` field and returns up to 10 matches
/// keyed by the trail's first vertex (good enough for "zoom to result").
public protocol TrailSearchRepository: Sendable {
    func searchTrails(_ query: String) async -> [SearchHit]
}

public struct GeonorgeTrailSearchRepository: TrailSearchRepository {
    private let session: URLSession
    private static let userAgent = "Turbo/0.1 github.com/SigmundGranaas/turbo"

    public init(session: URLSession = .shared) { self.session = session }

    public func searchTrails(_ query: String) async -> [SearchHit] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return [] }
        var components = URLComponents(string: "https://wfs.geonorge.no/skwms1/wfs.friluftsruter2")!
        let escaped = trimmed.replacingOccurrences(of: "'", with: "''")
        components.queryItems = [
            URLQueryItem(name: "SERVICE", value: "WFS"),
            URLQueryItem(name: "VERSION", value: "2.0.0"),
            URLQueryItem(name: "REQUEST", value: "GetFeature"),
            URLQueryItem(name: "TYPENAMES", value: "fotrute"),
            URLQueryItem(name: "OUTPUTFORMAT", value: "application/json"),
            URLQueryItem(name: "SRSNAME", value: "urn:ogc:def:crs:EPSG::4326"),
            URLQueryItem(name: "COUNT", value: "10"),
            URLQueryItem(name: "CQL_FILTER", value: "navn ILIKE '%\(escaped)%'"),
        ]
        guard let url = components.url else { return [] }
        var request = URLRequest(url: url)
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        guard let (data, response) = try? await session.data(for: request),
              (response as? HTTPURLResponse)?.statusCode == 200 else { return [] }
        return Self.parse(data)
    }

    // MARK: - Parsing (testable, network-free)

    static func parse(_ data: Data) -> [SearchHit] {
        guard let dto = try? JSONDecoder().decode(FeatureCollection.self, from: data) else { return [] }
        return dto.features.compactMap { feature in
            let name = feature.properties.navn ?? ""
            guard !name.isEmpty, let position = firstPoint(feature.geometry) else { return nil }
            let description = [feature.properties.rutenummer, feature.properties.merkemetode]
                .compactMap { $0 }.joined(separator: " · ")
            return SearchHit(name: name, description: description, position: position, kind: .hiking)
        }
    }

    static func firstPoint(_ geometry: Geometry) -> LatLng? {
        let pair: [Double]?
        switch geometry.type {
        case "LineString": pair = geometry.lineCoordinates?.first
        case "MultiLineString": pair = geometry.multiLineCoordinates?.first?.first
        default: pair = nil
        }
        guard let p = pair, p.count >= 2 else { return nil }
        return LatLng(lat: p[1], lng: p[0])   // GeoJSON is [lon, lat]
    }

    // MARK: - Wire types

    struct FeatureCollection: Decodable { let features: [Feature] }
    struct Feature: Decodable { let properties: Properties; let geometry: Geometry }
    struct Properties: Decodable {
        let navn: String?
        let rutenummer: String?
        let merkemetode: String?
    }
    struct Geometry: Decodable {
        let type: String
        let lineCoordinates: [[Double]]?
        let multiLineCoordinates: [[[Double]]]?

        enum CodingKeys: String, CodingKey { case type, coordinates }
        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            type = try c.decode(String.self, forKey: .type)
            // coordinates shape depends on geometry type — decode opportunistically.
            lineCoordinates = try? c.decode([[Double]].self, forKey: .coordinates)
            multiLineCoordinates = try? c.decode([[[Double]]].self, forKey: .coordinates)
        }
    }
}

public struct InMemoryTrailSearchRepository: TrailSearchRepository {
    private let hits: [SearchHit]
    public init(hits: [SearchHit] = []) { self.hits = hits }
    public func searchTrails(_ query: String) async -> [SearchHit] {
        let q = query.lowercased()
        return hits.filter { $0.name.lowercased().contains(q) }
    }
}

/// Merges multiple search sources behind one `SearchRepository`. Place results
/// (Kartverket stedsnavn) come first, then trail matches; recents delegate to the
/// primary place repository. Mirrors Flutter `CompositeSearchService`.
public final class CompositeSearchRepository: SearchRepository {
    private let place: SearchRepository
    private let trails: TrailSearchRepository

    public init(place: SearchRepository, trails: TrailSearchRepository) {
        self.place = place
        self.trails = trails
    }

    public func search(_ query: String) async -> [SearchHit] {
        async let placeHits = place.search(query)
        async let trailHits = trails.searchTrails(query)
        let (places, paths) = await (placeHits, trailHits)
        // De-dup trails whose name already appears in place results.
        let names = Set(places.map { $0.name.lowercased() })
        let freshTrails = paths.filter { !names.contains($0.name.lowercased()) }
        return places + freshTrails
    }

    public func recents() async -> AsyncStream<[RecentSearch]> { await place.recents() }
    public func addRecent(_ recent: RecentSearch) async { await place.addRecent(recent) }
}
